### PR TITLE
[#104] Infinite mass body collisions check

### PR DESCRIPTION
**Pull request by ilia-glushchenko**

### CODE REVIEW

File: Pegasus/include/pegasus/Collision.hpp [hunk 1]
```diff
@@ -110,19 +110,23 @@ class Detector
                 if (aObject.id == 0 || bObject.id == 0)
                     continue;
 
+                mechanics::Body& aBody = m_assetManager->GetAsset(m_assetManager->GetBodies(), aObject.data.body);
+                mechanics::Body& bBody = m_assetManager->GetAsset(m_assetManager->GetBodies(), bObject.data.body);
+                if (aBody.material.HasInfiniteMass() && bBody.material.HasInfiniteMass())
```
Comment (Warezovvv) [inline Pegasus/include/pegasus/Collision.hpp]: {
}

File: Pegasus/include/pegasus/Collision.hpp [hunk 2]
```diff
@@ -155,6 +159,11 @@ class Detector
...
+                    continue;
```
Comment (Warezovvv) [inline Pegasus/include/pegasus/Collision.hpp]: {
}

File: Pegasus/sources/Body.cpp [hunk 3]
```diff
@@ -32,6 +32,11 @@ void Body::Material::SetInfiniteMass()
     m_inverseMass = 0;
 }
 
+PEGASUS_EXPORT bool Body::Material::HasInfiniteMass() const
```
Comment (Warezovvv) [inline Pegasus/sources/Body.cpp]: PEGASUS_EXPORT in cpp

File: Pegasus/include/pegasus/Collision.hpp [hunk 4]
```diff
@@ -163,10 +172,7 @@ class Detector
                     && registeredContacts.find(key) == registeredContacts.end())
                 {
                     contacts.emplace_back(
-                        std::ref(m_assetManager->GetAsset(m_assetManager->GetBodies(), aObject.data.body)),
-                        std::ref(m_assetManager->GetAsset(m_assetManager->GetBodies(), bObject.data.body)),
-                        CalculateContactManifold(aShape, bShape),
-                        0.75
+                        std::ref(aBody), std::ref(bBody), CalculateContactManifold(aShape, bShape), 0.75
```
Comment (Warezovvv) [inline Pegasus/include/pegasus/Collision.hpp]: 0.75?

File: Pegasus/include/pegasus/Collision.hpp [hunk 5]
```diff
@@ -110,19 +110,23 @@ class Detector
...
 
+                mechanics::Body& aBody = m_assetManager->GetAsset(m_assetManager->GetBodies(), aObject.data.body);
+                mechanics::Body& bBody = m_assetManager->GetAsset(m_assetManager->GetBodies(), bObject.data.body);
+                if (aBody.material.HasInfiniteMass() && bBody.material.HasInfiniteMass())
+                    continue;
+
                 Shape* aShape = &m_assetManager->GetAsset(m_assetManager->GetShapes<Shape>(), aObject.data.shape);
                 Shape* bShape = &m_assetManager->GetAsset(m_assetManager->GetShapes<Shape>(), bObject.data.shape);
+                if (aShape == bShape)
+                    continue;
```
Comment (Warezovvv) [inline Pegasus/include/pegasus/Collision.hpp]: {
}

File: Pegasus/sources/Body.cpp (hunk 3 above)
Comment (dsodintsov) [inline Pegasus/sources/Body.cpp]: why macro in cpp?

Comment (ilia-glushchenko) [reply Pegasus/sources/Body.cpp]: Cause C++ ReSharper

**Pull request merged**